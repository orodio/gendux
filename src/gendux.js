import "babel-polyfill"

export const createStore = (reducer, state = {}) => {
  var subs = []

  const store = (function * store () {
    __loop: while (1) {
      const event = yield null
      if (event.type === "NOOP") continue __loop
      state = reducer(state, event)
      subs.forEach(fn => fn(state))
    }
  })()

  const dispatch = event => store.next(event)

  dispatch({ type:"NOOP" })

  const subscribe = fn => {
    subs.push(fn)
    var subscribed = true
    return () => {
      if (subscribed) subs.splice(subs.indexOf(fn), 1)
      subscribed = false
    }
  }

  return {
    dispatch,
    subscribe,
    getState: () => state
  }
}
