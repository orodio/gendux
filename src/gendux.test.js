import { createStore } from './gendux'

test('createStore', () => {
  const initState = { count:5 }
  const reducer = (state, event) => {
    if (event.type === "INC") return { ...state, count: state.count + 1 }
    return state
  }

  const mockReducer = jest.fn().mockImplementation(reducer)

  const { getState, dispatch, subscribe } = createStore(mockReducer, initState)

  expect(getState()).toMatchSnapshot()

  const event = { type:"INC" }

  dispatch(event)
  expect(mockReducer).toHaveBeenCalledWith(initState, event)
  const nextState = reducer(initState, event)
  expect(getState()).toMatchSnapshot()

  dispatch(event)
  expect(mockReducer).toHaveBeenCalledWith(nextState, event)
  expect(getState()).toMatchSnapshot()

  const sub = jest.fn()
  const unsub = subscribe(sub)

  dispatch(event)
  expect(getState()).toMatchSnapshot()

  dispatch(event)
  expect(getState()).toMatchSnapshot()

  unsub()

  dispatch(event)
  expect(getState()).toMatchSnapshot()

  dispatch(event)
  expect(getState()).toMatchSnapshot()

  expect(sub).toHaveBeenCalledTimes(2)
})
