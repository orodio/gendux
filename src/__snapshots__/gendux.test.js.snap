// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStore 1`] = `
Object {
  "count": 5,
}
`;

exports[`createStore 2`] = `
Object {
  "count": 6,
}
`;

exports[`createStore 3`] = `
Object {
  "count": 7,
}
`;

exports[`createStore 4`] = `
Object {
  "count": 8,
}
`;

exports[`createStore 5`] = `
Object {
  "count": 9,
}
`;

exports[`createStore 6`] = `
Object {
  "count": 10,
}
`;

exports[`createStore 7`] = `
Object {
  "count": 11,
}
`;
